index.d.ts

```diff
@@ -44,29 +44,40 @@ export type Operator = ComparisonOperator | CombinationOperator | NegationOperat
 export interface Filter extends Array<any> {}
 
 /**
- * A FunctionFilter that expects a string (propertyName) as second arguement and
- * a regular expression as third argument. An actual parser implementation has to
- * return a value for this function expression.
- */
-export interface StrMatchesFunctionFilter extends Filter {
-  0: 'FN_strMatches';
-  1: string;
-  2: RegExp;
```

Comment:
I think we should keep explicit function interfaces. This way the specific parsers can be developed closer to the style specification. A clompletly opened and unspecified interface adds more flexibility but be almost impossible to parse into all the different style representations.
IMHO extending the newly introduced `FunctionCall` but keeping the single interfaces would be the way to go.

Reply:
I'm not sure I agree on this. In SLD/SE for example this would be pretty much a one to one conversion. You can even use arbitrary expressions as values for Css/SvgParameters, not just for filter values. I'm not sure about this, but I think this also goes for at least the Mapbox styles (I don't know if they have support for functions though).

In order to map specific functions to style formats which only support some predefined functions you can still do a `switch` on the function name.

Reply:
Do you see any disadvantages in beeing precise for the functions? As we are trying to create a definition for a generalized style definition we should be as precise as we can.

Compare https://github.com/terrestris/geostyler-sld-parser/blob/master/src/SldStyleParser.ts#L212:
If the return type would be `FunctionCall` every array with a `string` at the first position and `array` at the second would be valid…

Reply:
I think the main disadvantage would be that we'd have to implement this for every function we want to support. Compare for example https://docs.geoserver.org/stable/en/user/filter/function_reference.html, in order to support all these we'd have to write a lot of code, while the generic approach would allow us to support them immediately.

I agree that the parsers would have to be rewritten at some places, but IMHO it would be worth it. Perhaps Lukas and I could approach this in the upcoming Bolsena Code Sprint.

Concerning the type safety, perhaps we should add a `type` tag to the expression types, that would also make it easier to identify them. What do you think?

Reply:
> I think the main disadvantage would be that we'd have to implement this for every function we want to support. Compare for example https://docs.geoserver.org/stable/en/user/filter/function_reference.html, in order to support all these we'd have to write a lot of code, while the generic approach would allow us to support them immediately.

It might be many lines of code, but its just a translation of the documentation. I think this can be done within a couple of hours. Maybe we should add extract the function expressions to a new file. But I still don't like the idea of being to dynamic here as we are creating a kind of a standard.

Maybe @jansule and @dnlkoch like to share some thoughts on this, too.

Reply:
I like the idea of being dynamic in here as our style specification would be very clear (IMHO) and we would support any custom or provider specific functions as well, but I see the point about providing a standarized and typed interface for the parser implementations and support the proposal of adding a named list of supported expression types.

Reply:
I also support the mindset of @KaiVolland to be as specific as possible. This will lead to much cleaner code in the parsers and makes parsing the expressions less prone to errors. Parsers will have to handle the different functions anyway, so there would not be too much overload.

We should also think about specifying the return types, which would be a huge advantage when thinking about nested expressions. Doing so might become a little tricky as we would need to be able to define return types for objects, not for functions...

Reply:
I still disagree for a simple reason: we would be able to express less than SLD/SE/FE. In the filter encoding standard it is possible to write any expression as operands of a filter. You can especially use a function call expression (`<ogc:Function name="sth">...`) anywhere, which can also have arbitrary expressions as operands.

The parsers don't need to know anything about those functions, they're not the ones evaluating them. They just need to know how to express a function call in the target language, and in SLD/SE it's just an `<ogc:Function>` element away. In CQL a function call is just `myFunc(arg1, arg2)`.

Reply:
> The parsers don't need to know anything about those functions, they're not the ones evaluating them. They just need to know how to express a function call in the target language, and in SLD/SE it's just an `<ogc:Function>` element away. In CQL a function call is just `myFunc(arg1, arg2)`.

This won't work for the [openlayers-parser](https://github.com/terrestris/geostyler-openlayers-parser):

If you have the function "sth" with arg1, arg2... SLD don't cares as the GeoServer knows what to do.

But there is no concept like function expressions in OpenLayers styles.The parser has to add it directly to the stylefunction.

Reply:
That's no different than for a predefined `strMatches` function. I'd say a function call expression would be translated in a style function like this:

```
// from the style definition:
const myFunction = {
  type: 'functioncall', // I'll add that later on
  name: 'callThisFunctionPlease',
  args: [{
    type: 'literal',
    value: 'some value'
  }, {
    type: 'property',
    name: 'some_property'
  }]
};

// the generated style function
const styleFunction = (feature) => {
  const transformedValue = callThisFunctionPlease('some value', feature.get('some_property');
  // proceed with transformedValue instead of a literal value
}
```

Reply:
But where is `callThisFunctionPlease` defined?

Reply:
I'm with you that it feels dirty, but just as GeoServer calls its function it would be the responsibility of the OpenLayers application in question to provide that function I guess.

In order to make it more clean we could have a convention to require people to provide the functions in some global namespace like `OlFunctions` to avoid cluttering `window`, or make this configurable in the OpenLayers-Parser, but ultimately the functions need to be defined by the target platform that evaluates and uses the style.

Or maybe someone else has another perspective on the story?